pintracker/stateless/stateless.go

```diff
@@ -0,0 +1,490 @@
+	statuses := spt.StatusAll()
+	resp := make([]api.PinInfo, 0)
+	for _, st := range statuses {
+		r, err := spt.Recover(st.Cid)
+		if err != nil {
+			return resp, err
+		}
+		resp = append(resp, r)
+	}
+	return resp, nil
+}
+
+// Recover will re-track or re-untrack a Cid in error state,
+// possibly retriggering an IPFS pinning operation and returning
+// only when it is done.
+func (spt *Tracker) Recover(c *cid.Cid) (api.PinInfo, error) {
+	logger.Infof("Attempting to recover %s", c)
+	pInfo, ok := spt.getError(c)
+	if !ok {
+		return spt.Status(c), nil
+	}
+
+	var err error
+	switch pInfo.Status {
+	case api.TrackerStatusPinError:
+		err = spt.enqueue(api.PinCid(c), optracker.OperationPin)
+	case api.TrackerStatusUnpinError:
+		err = spt.enqueue(api.PinCid(c), optracker.OperationUnpin)
+	}
+	if err != nil {
+		return spt.Status(c), err
+	}
+
+	pi, ok := spt.optracker.GetExists(c)
+	if !ok {
+		return spt.Status(c), nil
+	}
+
+	return pi, nil
+}
+
+func (spt *Tracker) ipfsStatusAll() (map[string]api.PinInfo, error) {
+	var ipsMap map[string]api.IPFSPinStatus
+	err := spt.rpcClient.Call(
+		"",
+		"Cluster",
+		"IPFSPinLs",
+		"recursive",
+		&ipsMap,
+	)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+	pins := make(map[string]api.PinInfo, 0)
+	for cidstr, ips := range ipsMap {
+		c, err := cid.Decode(cidstr)
+		if err != nil {
+			logger.Error(err)
+			continue
+		}
+		p := api.PinInfo{
+			Cid:    c,
+			Peer:   spt.peerID,
+			Status: ips.ToTrackerStatus(),
+		}
+		pins[cidstr] = p
+	}
+	return pins, nil
+}
+
+// localStatus returns a joint set of consensusState and ipfsStatus
+// marking pins which should be remote and leaving any ipfs pins that
+// aren't in the consensusState out.
+func (spt *Tracker) localStatus(incRemote bool) (map[string]api.PinInfo, error) {
+	pininfos := make(map[string]api.PinInfo)
+
+	// get shared state
+	var csps []api.PinSerial
+	err := spt.rpcClient.Call(
+		"",
+		"Cluster",
+		"Pins",
+		struct{}{},
+		&csps,
+	)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+	var cspis []api.Pin
+	for _, p := range csps {
+		cspis = append(cspis, p.ToPin())
+	}
+
+	// get statuses from ipfs node first
+	localpis, err := spt.ipfsStatusAll()
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	for _, p := range cspis {
+		if p.IsRemotePin(spt.peerID) && incRemote {
+			// add pin to pininfos with a status of remote
+			pininfos[p.Cid.String()] = api.PinInfo{
+				Cid:    p.Cid,
+				Peer:   spt.peerID,
+				Status: api.TrackerStatusRemote,
+			}
+			continue
+		}
+		// lookup p in localpis
+		if lp, ok := localpis[p.Cid.String()]; ok {
+			pininfos[p.Cid.String()] = lp
+		}
+	}
+	return pininfos, nil
+}
+
+func (spt *Tracker) getError(c *cid.Cid) (api.PinInfo, bool) {
+	return spt.optracker.FilterGet(c, optracker.PhaseError)
+}
+
+func (spt *Tracker) getErrorsAll() []api.PinInfo {
+	return spt.optracker.Filter(optracker.PhaseError)
+}
+
+func (spt *Tracker) removeError(c *cid.Cid) {
```

Comment:
this is not used. Also, `CleanError` should probably be used (atomic removal of error), but it can be called directly rather than having a wrapping method for the single call.

Reply:
✅